### DISCUSSION TITLE
Roadmap May 2023

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/llama.cpp#1220] [Announcements]
### High-prio

- [ ] **Refactoring pass**

  There is a lot of code duplication in `ggml.c` which probably can be simplified with a good set of macros. The goal is to keep the code size manageable, while we avoid reaching "macro hell"

- [x] **Optimize the AVX / AVX2 implementations of the quantization methods and add WASM SIMD**

  Make sure we have optimal implementation for these instruction sets

- [x] **Apply the new integer quantization methods to [whisper.cpp](https://github.com/ggerganov/whisper.cpp)**

  Will backport the latest `ggml` version to `whisper.cpp` and add support for quantized models. Will also update all WASM examples to be able to run with the quantized models

  **Update:** [whisper.cpp v1.4.0](https://github.com/ggerganov/whisper.cpp/releases/tag/v1.4.0) has been released. It includes integer quantization and GPU support via cuBLAS - all thanks to the great work done here

- [x] **Add support for "batch inference"**

  Recently, the [bert.cpp](https://github.com/skeskinen/bert.cpp) (by @skeskinen) project demonstrated BERT inference using `ggml`. This model gains a lot from batch inference, which is currently not supported by `ggml`. We will extend all operators to support it. The `bert.cpp` example will serve as a playground to achieve this

  **Update:** batched forward passes have been demonstrated in the [baby-llama](https://github.com/ggerganov/llama.cpp/blob/master/examples/baby-llama/baby-llama.cpp#L759-L770) example (thanks to @xaedes https://github.com/ggerganov/llama.cpp/pull/1360). It will be great to apply the demonstrated approach to `bert.cpp` and `whisper.cpp`'s beam-search decoding in order to gain extra speed-up

- [ ] **Implement inference of new models**

  There are already some very interesting models that should be supported by `ggml`:
  - [x] [💫 StarCoder](https://github.com/ggerganov/ggml/tree/master/examples/starcoder)
  - [ ] [Segment Anything Model (SAM)](https://github.com/facebookresearch/segment-anything)
  - [ ] [Bark (text-to-speech)](https://github.com/suno-ai/bark)
    There is a huge interest for adding `ggml` support for this model (see https://github.com/suno-ai/bark/issues/30#issuecomment-1526953640)
    The main blocker seems to be the dependency on [Facebook's EnCodec codec](https://github.com/facebookresearch/encodec). Still not sure how difficult it would be, but probably this codec is another model that we should try to support via `ggml`

  _I'll use this section to add a note regarding new model implementations by contributors - I recommend to always try to add a very basic example implementation to the [ggml](https://github.com/ggerganov/ggml/tree/master/examples) repo. Having a basic example there would make long-term support much easier_

- [x] **Proof-of-concept for 100% inference on the GPU**

  The goal is to make a demonstration of the idea discussed in #914 
  Very preliminary work has been started in https://github.com/ggerganov/ggml/pull/108
  Will try to get a working example using the [MNIST](https://github.com/ggerganov/ggml/tree/master/examples/mnist) inference

  **Update:** The MNIST inference on Apple Silicon GPU using Metal is now fully demonstrated: https://github.com/ggerganov/ggml/pull/108 -- this is the way

### Low-prio

- [ ] Project [ggml : improve threading implementation](https://github.com/users/ggerganov/projects/3)

  Better utilization of the available CPU resources via improved thread management.
  There have been a few efforts during April, but they remained in the "background" - need to put more focus this time

- [ ] **Having second thoughts about adding `llama_state`**

  The experience is we added `whisper_state` in `whisper.cpp` with this PR: https://github.com/ggerganov/whisper.cpp/pull/523

  However, I haven't see a lot of use of it. At the same time, it doubled the C API.
  Let me know if you think this is worth implementing

  ref: https://github.com/ggerganov/llama.cpp/pull/370#discussion_r1144027167

- [x] [Add 3-bit integer quantization](https://github.com/ggerganov/llama.cpp/issues/456)

  It has been shown that 2-bit integer quantization does not look really useful for anything: https://github.com/ggerganov/llama.cpp/pull/1004

  In this case, we can probably add 3-bit integer quantization. Probably not "officially supported", but rather in a state where we can run experiments and see if we can find some application

- [x] **There is an interesting ongoing effort to add "training" support to `ggml`:** https://github.com/ggerganov/ggml/issues/8#issuecomment-1519302173
 
  It would be really impressive if this actually works. Might be conflicts with the refactoring pass - need to coordinate with @xaedes 

  **Update:** this has been successfully completed and there is now a simple example demonstrating baby-LLaMA training: https://github.com/ggerganov/llama.cpp/blob/master/examples/baby-llama/baby-llama.cpp#L759-L770

**Comment by dakennedyd** [comment]


**Reply by verhovsky** [reply]
What do you mean?

**Reply by sroussey** [reply]
Likely meaning stablelm

**Reply by Green-Sky** [reply]
no stablelm is supported (by ggml, see repo) but its quality is underwhelming....
still, latent diffusion models would be sick

**Comment by 4t0m** [comment]
If I'm understanding the idea of llama_state--that it will allow multiple "inference threads" from a single loaded model, then it definitely seems worth implementing, since it opens up a lot of possibilities.

Is the idea that we can get a lot of the same gains by just quickly swapping out stored contexts?  A lot of llm applications benefit from having multiple instances that can build on one another, or different instances that receive diverse queries.

I haven't started using it yet, because I've been waiting for someone to post an example.  It's a bit hard for me to parse the api.

**Reply by FSSRepo** [reply]
The llama_state feature will be very useful when implementing an HTTP server because multiple clients can have multiple contexts at the same time. Currently, in an implementation I made of httplib in llama.cpp, it only supports a single client, meaning the context is unique.

**Reply by chrgeor** [reply]
I think there is another use case. I would break to multiple threads the next suggestion (maybe different seed) giving some options to the end user of the same context.

Sometimes there are not right wrong answers but more or less preferable

**Reply by maximegmd** [reply]
Wouldn't it be possible to change the llama_model in llama_context to a pointer so that we can assign models ourselves? Introduce a init_from_context function that takes another context and retrieves the llama_model from this context? 

As someone looking into using llama.cpp to run multi step agents I would really appreciate being able to have multiple contexts sharing the same model.

**Reply by ggerganov** [reply]
The right way to do it is like we do it in `whisper.cpp`:

https://github.com/ggerganov/whisper.cpp/blob/master/whisper.h

If someone wants to give it a try at implementing it here

**Reply by bullno1** [reply]
Should every public API has an implicit and an explicit `_with_state` & `_from_state` version though? That seems pretty verbose.
Granted, the implicit version could probably just forward to the explicit version.

Still, does that mean if one choose the explicit state version, the context would still contain an unused state?

**Reply by ggerganov** [reply]
I agree that it is a bit over-verbose, but we didn't see a better way. Open to suggestions

> Still, does that mean if one choose the explicit state version, the context would still contain an unused state?

There are init calls that explicitly do not create an internal state:

https://github.com/ggerganov/whisper.cpp/blob/57543c169e27312e7546d07ed0d8c6eb806ebc36/whisper.h#L109

**Reply by didzis** [reply]
I suggest using the `llama_context` as the `llama_state`. For me that seems semantically correct and the most logical approach, it is fully backwards compatible requiring no changes for existing public API users. It's also very simple with only few changes internally and only few additions to the public API only for those who would like to use this feature.

I created a pull request here: https://github.com/ggerganov/llama.cpp/pull/1797#issue-1751200897

**Reply by bullno1** [reply]
I just tried to create a `llama_state` but the duplication in sampling is too much.
All sampling functions rely on `ctx->rng`.

It's also a constant duplication factor to all new public API going forward.

@didzis change is great, just 2 new public functions.

**Comment by monircsueb** [comment]
do we have weigh model support for Encodec? If yes, could you please tell me how to build the sgml-model.bin? 
Thank you appreciate your help and time.
